### PR TITLE
Implement Tooltip

### PR DESCRIPTION
**Pull request by eddie0329**
issue: [missing tooltip](https://github.com/bestguy/sveltestrap/issues/31)

Hi @bestguy. I implement tooltip just as we've discussed at [issue](https://github.com/bestguy/sveltestrap/issues/31). I think there are no proper template for PRs. So i write it with personal ones. 

Advices and comments are welcome. 

## 📌  Works

- Implementing tooltip
- unit test tooltip
- Add story tooltip
- Install popperjs

## 📌  Discussion

Currently i limited placement into 4, which is top, left, right and bottom. 

Should i extend to 12 different sections as , left-start, left, left-end..etc ? Or should i stick to current.

### CODE REVIEW

**Comment by bestguy** [comment]
Ah thank you @eddie0329 !  Very appreciated figuring the popper usage out.  
I think it's acceptable to just use the 4 directions, as [Bootstrap docs](https://getbootstrap.com/docs/4.5/components/tooltips/) only mention the 4.

I'll try and review asap!

File: src/Tooltip.svelte [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    isHover ? 'show' : false
+  );
+</script>
+
+<div
+  id={tooltipId}
+  class={classes}
+  role="tooltip"
+  x-placement={popperPlacement}
+  aria-label={ariaLabel}>
```
Comment (bestguy) [inline src/Tooltip.svelte:69]: You can simplify this to:

```html
<div
  bind:this={tooltipEl}
  {...$$restProps}
  class={classes}
  role="tooltip"
  x-placement={popperPlacement}>
```


And remove the `tooltipId`, `generateRandomDigits` usage

File: src/utils.js [hunk 2]
```diff
@@ -107,3 +107,9 @@ function toClassName(value) {
 export default function classnames(...args) {
   return args.map(toClassName).filter(Boolean).join(' ');
 }
+
+export const generateRandomDigits = (argMin, argMax) => {
+  const min = Math.ceil(argMin);
+  const max = Math.floor(argMax);
+  return Math.floor(Math.random() * (max - min)) + min;
+};
```
Comment (bestguy) [inline src/utils.js:115]: I think you can remove this as `bind:this={tooltipEl}` above will take care of this.

File: src/Tooltip.svelte [hunk 3]
```diff
@@ -0,0 +1,78 @@
+<script>
+  import { onMount } from 'svelte';
+  import { createPopper } from '@popperjs/core';
+  import classnames, { generateRandomDigits } from './utils';
```
Comment (bestguy) [inline src/Tooltip.svelte:4]: See below

File: src/Tooltip.svelte [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  export let target = '';
+  export let placement = 'top';
+  export let children = undefined;
+  let isHover = false;
+  let popperInstance;
+  let popperPlacement = placement;
+  let tooltipEl;
+  let targetEl;
+
+  const tooltipId = `tooltip${generateRandomDigits(0, 100000)}`;
```
Comment (bestguy) [inline src/Tooltip.svelte:17]: See below

File: src/Tooltip.svelte [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+  const disableHover = () => {
+    isHover = false;
+    if (popperInstance) {
+      popperInstance.destory();
+    }
+  };
+
+  onMount(() => {
+    tooltipEl = document.querySelector(`#${tooltipId}`);
```
Comment (bestguy) [inline src/Tooltip.svelte:43]: See below

File: src/Tooltip.svelte [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    targetEl = document.querySelector(`#${target}`);
+    targetEl.addEventListener('mouseover', enableHover);
+    targetEl.addEventListener('mouseleave', disableHover);
+    popperInstance = createPopper(targetEl, tooltipEl, {
+      placement,
+      modifiers: [checkPopperPlacement]
+    });
+  });
+
+  $: ariaLabel = $$props['aria-label'];
```
Comment (bestguy) [inline src/Tooltip.svelte:53]: See below

**Review by bestguy — CHANGES_REQUESTED** [review]
This looks great, thank you. One suggested change to simplify.

**Comment by bestguy** [comment]
Also, you need to add:

```javascript
export { default as Toast } from './Toast.svelte';
```

from index.js

File: src/Tooltip.svelte (hunk 1 above)
Comment (eddie0329) [reply src/Tooltip.svelte:69]: oh Thank you for comment ! i will change this as you mentioned

**Comment by eddie0329** [comment]
> This looks great, thank you. One suggested change to simplify.

Reallly appreciate about reviews! I will change these. Again thank you.

**Comment by eddie0329** [comment]
@bestguy I changed code with following:

- bind:this={tooltipEl} work
- {...$$restProps} work for like aria-label
- remove unuesd codes due to updating code review
- Add tooltip in index.js file

**Review by bestguy — APPROVED** [review]
Wonderful!   Thank you for the help.

**Pull request merged**